File: problem4/java-solution/P4.java
/*
 * P3.java
 *
 * Zeke Reyna
 *
 * Problem 4:
 * Find the largest palindrome made from the product of two 3-digit numbers.
 */

public class P4
{
        public static int set_lower(int digits)
        {
                int lower = 1;
                for (int i = 1; i < digits; i++)
                        lower *= 10;
                return lower;
        }

        public static int set_upper(int digits)
        {
                int upper = 1;
                for (int i = 0; i < digits; i++)
                        upper *= 10;
                return upper;
        }

        public static int check_if_prod_is_pal(int first_term, int second_term)
        {
                int product, flipped, temp;
                product = first_term * second_term;
                flipped = 0;
                temp = product;

                while (temp > 0) {
                        flipped *= 10;
                        flipped += temp % 10;
                        temp /= 10;
                }

                if (product != flipped)
                        product = 0;
                return product;
        }

        public static int find_largest_pal(int lower, int upper)
        {
                int i, j;
                int cur_pal;
                int largest_pal = 0;
                for (i = lower; i < upper; i++) {
                        for (j = i; j < upper; j++) {
                                cur_pal = check_if_prod_is_pal(i, j);
                                if (cur_pal > largest_pal)
                                        largest_pal = cur_pal;
                        }
                }
                return largest_pal;
        }

        public static void main(String args[])
        {
                int largest_pal = 0;
                int num_digits;
                if (args.length == 1)
                        num_digits = Integer.parseInt(args[0]);
                else
                        num_digits = 3; // default

                int lower_bound, upper_bound;
                lower_bound = set_lower(num_digits);
                upper_bound = set_upper(num_digits);

                largest_pal = find_largest_pal(lower_bound, upper_bound);

                System.out.print("\nThe largest palindrome that is made from the product of");
                System.out.println(" two "+num_digits+"-digit numbers is: "+largest_pal+"\n");

        }
}

File: problem4/cpp-solution/p4.cpp
/*
 * p4.cpp
 *
 * Zeke Reyna
 *
 * Problem 4:
 * Find the largest palindrome made from the product of
 * two 3-digit numbers.
 *
 * note: solves for any n-digit number, where n resides in the bounds of 'int'
 */

#include <iostream>
using namespace std;
const int _TEST = 2;
const int _DEFAULT = 3;

void set_bounds(int& lower, int& upper, const int digits)
{
        lower = 1;
        upper = 1;
        for (int i = 1; i < digits; i++)
                lower *= 10;
        for (int i = 0; i < digits; i++)
                upper *= 10;
}

int check_if_prod_is_pal(const int first_term, const int second_term)
{
        int product = first_term * second_term;
        int flipped = 0;
        int temp = product;

        while (temp > 0) {
                flipped *= 10;
                flipped += temp % 10;
                temp /= 10;
        }

        if (product != flipped)
                product = 0;
        return product;
}

int find_largest_pal(const int lower, const int upper)
{
        int largest_pal = 0;
        for (int i = lower; i < upper; i++) {
                for (int j = i; j < upper; j++) {
                        int cur_pal = check_if_prod_is_pal(i, j);
                        if (cur_pal > largest_pal)
                                largest_pal = cur_pal;
                }
        }
        return largest_pal;
}

int main(int argc, char **argv)
{
	ios_base::sync_with_stdio(false);
        int num_digits;
        (argc == 2) ? (num_digits = stoi(argv[1])) : (num_digits = _DEFAULT);

        int lower_bound, upper_bound;
        set_bounds(lower_bound, upper_bound, num_digits);

        int largest_pal = find_largest_pal(lower_bound, upper_bound);

        cout<<"\nThe largest palindrome that is made from the product of";
        cout<<" two "<<num_digits<<"-digit numbers is: "<<largest_pal<<"\n\n";
        return 0;
}
